### PR TITLE
feat(webapp): add tracing page

### PR DESCRIPTION
**Pull request by dogfrogfog**
## Brief
 - docs.google
## Changes
- add /exemplars/single page
- rename /tracing -> /exemplars/merge
- add heatmap component
- add selection to heatmap compoent
- fetch flamegraph with heatmap selected area values

## Concerns
-  resize bug

### CODE REVIEW

**Comment by dogfrogfog** [comment]
https://user-images.githubusercontent.com/47758224/186949537-2c308c1f-fbef-4019-8474-917c4288f3e8.mp4

**Comment by dogfrogfog** [comment]
[ref](https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html)
cc @Rperry2174 @eh-am 

Screenshot 2022-09-05 at 17 43 56

File: webapp/javascript/pages/Heatmap/Heatmap.tsx [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+          backgroundImage: `linear-gradient(to right, ${Color.rgb(
+            COLOR_1
+          )} , ${Color.rgb(COLOR_2)})`,
+        }}
+      >
+        <span role="textbox" style={{ color: Color.rgb(COLOR_2).toString() }}>
+          {minValue}
+        </span>
+        <span role="textbox" style={{ color: Color.rgb(COLOR_1).toString() }}>
+          {maxValue}
```
Comment (petethepig) [inline webapp/javascript/pages/Heatmap/Heatmap.tsx:158]: This should be Z value instead of Y value.

**Comment by dogfrogfog** [comment]
https://user-images.githubusercontent.com/47758224/189398527-d394b3bc-1d31-4278-9ec0-cf8cebc78f42.mov

File: webapp/javascript/services/render.ts [hunk 2]
```diff
@@ -70,6 +70,7 @@ interface mergeWithQueryIDProps {
   maxNodes: string | number;
 }
 
+// z.infer<typeof MergeMetadataSchema> ?
```
Comment (Rperry2174) [inline webapp/javascript/services/render.ts:73]: do we still need this here

File: webapp/javascript/services/render.ts [hunk 3]
```diff
@@ -126,6 +127,123 @@ export async function mergeWithQueryID(
...
+  profile: Profile;
+}
+
+export async function getHeatmap(
+  props: getHeatmapProps,
+  controller?: {
+    signal?: AbortSignal;
+  }
+): Promise<Result<HeatmapOutput, RequestError | ZodError>> {
+  // todo use common builder
```
Comment (Rperry2174) [inline webapp/javascript/services/render.ts:165]: If you're going to leave this comment here can you elaborate on what you mean by this?

File: webapp/javascript/components/Heatmap/Heatmap.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    expect(xAxisTicks).toHaveLength(6);
+  });
+
+  it('should have correct color scale', () => {
+    render(<Heatmap />);
+
+    const [minTextEl, maxTextEl] = within(
+      screen.getByTestId('color-scale')
+    ).getAllByRole('textbox');
+    // expect(minTextEl.textContent).toBe(`${exemplarsQueryHeatmap.minDepth - 1}`);
```
Comment (Rperry2174) [inline webapp/javascript/components/Heatmap/Heatmap.spec.tsx:53]: Do we need these here?

File: webapp/javascript/components/Heatmap/Heatmap.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+jest.mock('./useHeatmapSelection.hook', () => ({
+  ...jest.requireActual('./useHeatmapSelection.hook'),
+  useHeatmapSelection: () => ({
+    selectedCoordinates: { start: null, end: null },
+    selectedAreaToHeatmapRatio: 1,
+    hasSelectedArea: false,
+  }),
+}));
+
+// TODO(dogfrogfog): refactor
```
Comment (Rperry2174) [inline webapp/javascript/components/Heatmap/Heatmap.spec.tsx:16]: refactor in what what?

File: webapp/javascript/components/Heatmap/Heatmap.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+describe.skip('Component: Heatmap', () => {
```
Comment (eh-am) [inline webapp/javascript/components/Heatmap/Heatmap.spec.tsx:17]: shouldn't be skipped

File: webapp/javascript/services/render.ts [hunk 7]
```diff
@@ -126,6 +127,123 @@ export async function mergeWithQueryID(
...
+  props: selectionProfileProps,
+  controller?: {
+    signal?: AbortSignal;
+  }
+): Promise<Result<SelectionProfileOutput, RequestError | ZodError>> {
+  const queryString = Object.entries(props).reduce(
+    (acc, [key, value]) => acc + (acc ? `&${key}=${value}` : `${key}=${value}`),
+    ''
+  );
+  const response = await request(`/api/exemplars:query?${queryString}`, {
```
Comment (eh-am) [inline webapp/javascript/services/render.ts:224]: @kolesnikovae I am curious about the usage of `:` in a url?

File: webapp/javascript/services/render.ts [hunk 8]
```diff
@@ -126,6 +127,123 @@ export async function mergeWithQueryID(
...
+
+export async function getHeatmapSelectionProfile(
+  props: selectionProfileProps,
+  controller?: {
+    signal?: AbortSignal;
+  }
+): Promise<Result<SelectionProfileOutput, RequestError | ZodError>> {
+  const queryString = Object.entries(props).reduce(
+    (acc, [key, value]) => acc + (acc ? `&${key}=${value}` : `${key}=${value}`),
+    ''
```
Comment (eh-am) [inline webapp/javascript/services/render.ts:222]: Prefer to use `URLSearchParams`

File: webapp/javascript/services/exemplarsTestData.ts [hunk 9]
```diff
@@ -0,0 +1,252 @@
+export const heatmapMockData = {
```
Comment (eh-am) [inline webapp/javascript/services/exemplarsTestData.ts:1]: I guess this file doesn't have to exist anymore?

File: webapp/javascript/redux/reducers/tracing.ts [hunk 10]
```diff
@@ -151,6 +241,128 @@ export const tracingSlice = createSlice({
...
+          state.heatmapSingleView = { type: 'loading' };
+        }
+      }
+    });
+
+    builder.addCase(fetchSelectionProfile.fulfilled, (state, action) => {
+      state.heatmapSingleView = {
+        ...action.payload,
+        // refactor types
+        heatmap: state.heatmapSingleView.heatmap as Heatmap,
```
Comment (eh-am) [inline webapp/javascript/redux/reducers/tracing.ts:332]: Why is this assertion needed?

File: webapp/javascript/redux/reducers/tracing.ts [hunk 11]
```diff
@@ -79,6 +98,77 @@ export const fetchSingleView = createAsyncThunk<
...
+  }
+
+  if (res.isErr && res.error instanceof RequestAbortedError) {
+    return thunkAPI.rejectWithValue({ rejectedWithValue: 'reloading' });
+  }
+
+  thunkAPI.dispatch(
+    addNotification({
+      type: 'danger',
+      title: 'Failed to load single view data',
```
Comment (eh-am) [inline webapp/javascript/redux/reducers/tracing.ts:164]:
```suggestion
      title: 'Failed to load profile',
```

File: webapp/javascript/redux/reducers/tracing.ts [hunk 12]
```diff
@@ -79,6 +98,77 @@ export const fetchSingleView = createAsyncThunk<
...
```
Comment (eh-am) [inline webapp/javascript/redux/reducers/tracing.ts:127]:
```suggestion
      title: 'Failed to load heatmap',
```

File: webapp/javascript/components/Heatmap/useHeatmapSelection.hook.ts [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+          query,
+          from,
+          until,
+          ...DEFAULT_HEATMAP_PARAMS,
+        })
+      );
+      return () => fetchData.abort('cancel');
+    }
+    return undefined;
+  }, [from, until, query]);
```
Comment (eh-am) [inline webapp/javascript/components/Heatmap/useHeatmapSelection.hook.ts:231]: I think this is waaaay too much functionality for this single hook. Let's move data fetching to the component/another hook.

**Review by eh-am — COMMENTED** [review]
Wow, Looks good and quite impressive!!

I would like to see the boundaries more clear though:
A generic `Heatmap` that you pass data and callback events. Right now it's too intertwined with tracing stuff. As an illustration, the `HeatMap` component should know nothing about the redux store or its state. It should just receive `heatmapData` (in whatever format it has) and its callbacks (onSelection I guess?).

File: webapp/javascript/services/exemplarsTestData.ts (hunk 9 above)
Comment (dogfrogfog) [reply webapp/javascript/services/exemplarsTestData.ts:1]: we can use it to mock `/api/exemplars` response

**Review by Rperry2174 — APPROVED** [review]
LGTM! Nice work @dogfrogfog

**Pull request merged**